### PR TITLE
Feat/spec tranformer

### PR DESCRIPTION
**Pull request by zamhown**
[[中文版模板 / Chinese template](https://github.com/VisActor/VChart/blob/main/.github/PULL_REQUEST_TEMPLATE/pr_cn.md?plain=1)]

### 🤔 This is a ...

- [ ] New feature
- [ ] Bug fix
- [ ] TypeScript definition update
- [ ] Bundle size optimization
- [ ] Performance optimization
- [ ] Enhancement feature
- [x] Refactoring
- [ ] Update dependency
- [ ] Code style optimization
- [ ] Test Case
- [ ] Branch merge
- [ ] Release
- [ ] Site / documentation update
- [ ] Demo update
- [ ] Workflow
- [ ] Other (about what?)

### 🔗 Related issue link
#1725 #1729 

### 📝 Changelog

| Language   | Changelog |
| ---------- | --------- |
| 🇺🇸 English |           |
| 🇨🇳 Chinese |           |

### ☑️ Self-Check before Merge

⚠️ Please check all items below before requesting a reviewing. ⚠️

- [ ] Doc is updated/provided or not needed
- [ ] Demo is updated/provided or not needed
- [ ] TypeScript definition is updated/provided or not needed
- [ ] Changelog is provided or not needed

---

### 🚀 Summary

copilot:summary

### 🔍 Walkthrough

copilot:walkthrough

### CODE REVIEW

File: packages/vchart/src/core/vchart.ts [hunk 1]
```diff
@@ -371,12 +387,41 @@ export class VChart implements IVChart {
...
+      spec = JSON.parse(spec);
+    }
+    if (forceMerge && this._spec) {
+      spec = mergeSpec({}, this._spec, spec);
+    }
+    spec = specTransform(spec) as any;
+    // because of in data-init, data will be set as array;
+    spec.data = spec.data ?? [];
+
+    this._originalSpec = cloneDeepSpec(spec);
```
Comment (xile611) [inline packages/vchart/src/core/vchart.ts:405]: 按照逻辑是不是应该 `this._originalSpec = spec;`
`this._spec =  cloneDeepSpec(spec)`;

File: packages/vchart/src/core/vchart.ts [hunk 2]
```diff
@@ -371,12 +387,41 @@ export class VChart implements IVChart {
...
+  /** 设置新 spec，返回是否成功 */
+  private _setNewSpec(spec: any, forceMerge?: boolean): boolean {
     if (!spec) {
-      return;
+      return false;
+    }
+    if (isString(spec)) {
+      spec = JSON.parse(spec);
+    }
+    if (forceMerge && this._spec) {
```
Comment (xile611) [inline packages/vchart/src/core/vchart.ts:398]: `_originalSpec ` 的赋值是不是应该在 merge 和 `specTranfrom` 之前？

File: packages/vchart/src/core/vchart.ts [hunk 3]
```diff
@@ -371,12 +387,41 @@ export class VChart implements IVChart {
...
+      spec = mergeSpec({}, this._spec, spec);
+    }
+    spec = specTransform(spec) as any;
```
Comment (xile611) [inline packages/vchart/src/core/vchart.ts:401]: 这个 `specTranform` 本身相当于做了`deepClone`，只是不处理遍历data，返回结果 和 `spec` 指向不同引用，下面的 `cloneDeepSpec`也没必要了吧

File: packages/vchart/src/component/data-zoom/data-zoom/data-zoom.ts [hunk 4]
```diff
@@ -40,19 +49,38 @@ export class DataZoom<T extends IDataZoomSpec = IDataZoomSpec> extends DataFilte
...
+        spec: s,
+        specIndex: i,
+        specPath: [this.specKey, i],
+        type: ComponentTypeEnum.dataZoom
+      });
+    });
+    return specInfos;
+  }
+
+  static createComponent(specInfo: IModelSpecInfo, options: IComponentOption) {
```
Comment (kkxxkk2019) [inline packages/vchart/src/component/data-zoom/data-zoom/data-zoom.ts:78]: 这个静态方法是不是可以不用每个组件类都实现一遍了？

File: packages/vchart/src/core/vchart.ts [hunk 5]
```diff
@@ -259,10 +265,20 @@ export class VChart implements IVChart {
   /** 工具方法 */
   static readonly Utils = VCHART_UTILS;
 
+  protected _originalSpec: any;
   protected _spec: any;
+  getSpec() {
+    return this._spec;
```
Comment (kkxxkk2019) [inline packages/vchart/src/core/vchart.ts:271]: 这个方法是返回用户的 spec 还是内部的？

File: packages/vchart/src/core/vchart.ts (hunk 1 above)
Comment (zamhown) [reply packages/vchart/src/core/vchart.ts:405]: 嗯嗯，我改了

File: packages/vchart/src/core/vchart.ts (hunk 2 above)
Comment (zamhown) [reply packages/vchart/src/core/vchart.ts:398]: 这样的话设置新主题的时候会重新走一遍 merge 和 specTranfrom 流程

File: packages/vchart/src/core/vchart.ts (hunk 3 above)
Comment (zamhown) [reply packages/vchart/src/core/vchart.ts:401]: 这个确实，那我在这里做一个 spec 和 originalSpec 的分隔？不过后续用到 originalSpec 时都要重新 specTranform 一下

File: packages/vchart/src/component/data-zoom/data-zoom/data-zoom.ts (hunk 4 above)
Comment (zamhown) [reply packages/vchart/src/component/data-zoom/data-zoom/data-zoom.ts:78]: 是的，我删了，只剩 axis

File: packages/vchart/src/core/vchart.ts (hunk 5 above)
Comment (zamhown) [reply packages/vchart/src/core/vchart.ts:271]: 返回内部的

**Pull request merged**